### PR TITLE
refactor(interchain-token-service): create message execution interceptors

### PR DESCRIPTION
**Pull request by fish-sammy**
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: contracts/interchain-token-service/src/state.rs
Comment (cgorenflo) [inline contracts/interchain-token-service/src/state.rs:1]: this change doesn't make sense to me

File: contracts/interchain-token-service/src/contract/execute/mod.rs [hunk 1]
```diff
@@ -150,17 +150,61 @@ fn apply_to_hub(
...
+        .and_then(|_| {
+            interceptors::apply_scaling_factor_to_amount(
+                storage,
+                &source_chain,
+                &destination_chain,
+                &mut transfer,
+            )
+        })
+        .and_then(|_| interceptors::add_supply_amount(storage, &destination_chain, &transfer))
+        .map(|_| transfer)
```
Comment (cgorenflo) [inline contracts/interchain-token-service/src/contract/execute/mod.rs:180]: not that I don't like more functional patterns, but is this necessary here?
Wouldn't it just read
```
interceptors::subtract_supply_amount(...)?;
interceptors::apply_scaling_factor_to_amount(...)?;
interceptors::add_supply_amount(...)?;

Ok(transfer);
```

File: contracts/interchain-token-service/src/contract/execute/mod.rs [hunk 2]
```diff
@@ -150,17 +150,61 @@ fn apply_to_hub(
...
+            )
+        })
+        .and_then(|_| {
+            interceptors::save_token_instance_for_destination_chain(
+                storage,
+                destination_chain,
+                &deploy_token,
+            )
+        })
+        .map(|_| deploy_token)
```
Comment (cgorenflo) [inline contracts/interchain-token-service/src/contract/execute/mod.rs:205]: same question

File: contracts/interchain-token-service/src/state.rs
Comment (fish-sammy) [reply contracts/interchain-token-service/src/state.rs:1]: TokenId implements Copy so it feels extra to me to pass it by reference, but not a very strong opinion.

File: contracts/interchain-token-service/src/contract/execute/mod.rs (hunk 1 above)
Comment (fish-sammy) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:180]: IMO, leaving it functional would encourage people to do one thing at a step in the future when they need to touch this function again. What I want to avoid for this function to happen is people start adding procedural things to it again.

Comment (cgorenflo) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:180]: would it be possible for these functions to return the appropriate message, i.e. take in `transfer` and pass it back out? That way, the `and_then` wouldn't be wasted and you wouldn't need to map in the end

Comment (fish-sammy) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:180]: Thought of doing it, but decided to land like this so that it's clearer which interceptors would mutate the messages and which ones would not when someone looks at the apply function.

Comment (cgorenflo) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:180]: It's a good argument, but is that a concern?

File: contracts/interchain-token-service/src/state.rs
Comment (cgorenflo) [reply contracts/interchain-token-service/src/state.rs:1]: ah gotcha

File: contracts/interchain-token-service/src/contract/execute/mod.rs (hunk 1 above)
Comment (fish-sammy) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:180]: No, it's just better readability.

Comment (cjcobb23) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:180]: I actually disagree, I think the mutability here is a bit hard to follow. I think it would be better if `apply_scaling_factor_to_amount` actually returned the new transfer, instead of mutating. Same with the other functions. We can then save the returned values in this function (if necessary).

File: contracts/interchain-token-service/src/contract/execute/interceptors.rs [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+
+    state::save_token_instance(storage, chain.clone(), transfer.token_id, &token)
+        .change_context(Error::State)
+}
+
+pub fn apply_scaling_factor_to_amount(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    destination_chain: &ChainNameRaw,
+    transfer: &mut InterchainTransfer,
```
Comment (cjcobb23) [inline contracts/interchain-token-service/src/contract/execute/interceptors.rs:52]: just return this, instead of mutating. The function doesn't return anything as is

File: contracts/interchain-token-service/src/contract/execute/interceptors.rs [hunk 4]
```diff
@@ -0,0 +1,648 @@
...
+
+use super::Error;
+use crate::state::{self, TokenDeploymentType};
+use crate::{DeployInterchainToken, InterchainTransfer, TokenConfig, TokenId, TokenInstance};
+
+pub fn subtract_supply_amount(
+    storage: &mut dyn Storage,
+    chain: &ChainNameRaw,
+    transfer: &InterchainTransfer,
+) -> Result<(), Error> {
```
Comment (cjcobb23) [inline contracts/interchain-token-service/src/contract/execute/interceptors.rs:14]: I think this would read better if you passed in the `TokenInstance`, subtracted the amount, and returned the new instance. And then the caller can save. I don't think the name and signature of this function gives a good idea of what it's doing

File: contracts/interchain-token-service/src/contract/execute/interceptors.rs [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+
+    state::save_token_instance(storage, chain.clone(), transfer.token_id, &token)
+        .change_context(Error::State)
+}
+
+pub fn add_supply_amount(
+    storage: &mut dyn Storage,
+    chain: &ChainNameRaw,
+    transfer: &InterchainTransfer,
+) -> Result<(), Error> {
```
Comment (cjcobb23) [inline contracts/interchain-token-service/src/contract/execute/interceptors.rs:33]: Same comment as above

File: contracts/interchain-token-service/src/contract/execute/interceptors.rs [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+        source_chain,
+        destination_chain,
+        transfer.token_id,
+        transfer.amount,
+    )?;
+
+    Ok(())
+}
+
+pub fn save_token_instance_for_source_chain(
```
Comment (cjcobb23) [inline contracts/interchain-token-service/src/contract/execute/interceptors.rs:65]: I don't think this is a good name for this function, I generally think functions that are called `save_something`, just save something. But actually this is doing quite a bit of initialization and checks.

File: contracts/interchain-token-service/src/contract/execute/interceptors.rs [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+    )
+    .change_context(Error::State)
+    .map(|_| ())
+}
+
+pub fn calculate_scaling_factor(
+    storage: &dyn Storage,
+    source_chain: &ChainNameRaw,
+    destination_chain: &ChainNameRaw,
+    deploy_token: &mut DeployInterchainToken,
```
Comment (cjcobb23) [inline contracts/interchain-token-service/src/contract/execute/interceptors.rs:132]: why not just return this? This mutability is hard to read.

**Review by cjcobb23 — CHANGES_REQUESTED** [review]
I dislike the mutation of function arguments that's in this change, it feels like a bit of an anti pattern. I'm happy to be wrong, but would like to discuss with the team before merging. I don't think we do this anywhere else in the code.

File: contracts/interchain-token-service/src/contract/execute/mod.rs (hunk 1 above)
Comment (fish-sammy) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:180]: Returning it hides the fact that it may have changed during the process. Someone looking at the caller's flow can not tell which interceptors transform the message and which do not. It's not an anti-pattern. Rust forces people to use the `mut` keyword for a reason.

File: contracts/interchain-token-service/src/contract/execute/interceptors.rs (hunk 6 above)
Comment (fish-sammy) [reply contracts/interchain-token-service/src/contract/execute/interceptors.rs:65]: What do u suggest?

File: contracts/interchain-token-service/src/contract/execute/mod.rs [hunk 8]
```diff
@@ -151,20 +150,56 @@ fn apply_to_hub(
...
+    interceptors::subtract_supply_amount(storage, &source_chain, &transfer)?;
+    let transfer = interceptors::apply_scaling_factor_to_amount(
+        storage,
+        &source_chain,
+        &destination_chain,
+        transfer,
+    )?;
+    interceptors::add_supply_amount(storage, &destination_chain, &transfer)?;
+
+    Ok(transfer)
```
Comment (cgorenflo) [inline contracts/interchain-token-service/src/contract/execute/mod.rs:179]: here `and_then` might look better now 😆

File: contracts/interchain-token-service/src/contract/execute/mod.rs [hunk 9]
```diff
@@ -151,20 +150,56 @@ fn apply_to_hub(
...
+    interceptors::deploy_token_to_source_chain(storage, source_chain, &deploy_token)?;
+    let deploy_token = interceptors::calculate_scaling_factor(
+        storage,
+        source_chain,
+        destination_chain,
+        deploy_token,
+    )?;
+    interceptors::deploy_token_to_destination_chain(storage, destination_chain, &deploy_token)?;
+
+    Ok(deploy_token)
```
Comment (cgorenflo) [inline contracts/interchain-token-service/src/contract/execute/mod.rs:197]: same

File: contracts/interchain-token-service/src/contract/execute/mod.rs (hunk 8 above)
Comment (cgorenflo) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:179]: my biggest issue before was that it was just used to unpack the result, but the passed value was ignored. With the new fluid interface that is no longer the case

Comment (cgorenflo) [reply contracts/interchain-token-service/src/contract/execute/mod.rs:179]:
```suggestion
    interceptors::apply_scaling_factor_to_amount(
        storage,
        &source_chain,
        &destination_chain,
        transfer,
    )
    .inspect(interceptors::add_supply_amount(storage, &destination_chain, transfer))
```

**Review by cjcobb23 — APPROVED** [review]
